effekt/shared/src/main/scala/effekt/Typer.scala

```diff
@@ -311,7 +311,10 @@ object Typer extends Phase[NameResolved, Typechecked] {
         Context.error("Duplicate definitions of operations")
 
       clauses foreach Context.withFocus {
-        case d @ source.OpClause(op, tparams, params, body, resume) =>
+        case _ @ source.OpClause(op, tparams, params, Some(ret), body, resumeId) =>
+          // TODO: Can we somehow use the return type provided by the user?
+          Context.abort(pretty"Unexpected type annotation on operation ${op}.")
```

Comment:
Does "use" here mean using it for error messages, or for type checking?

Reply:
To be precise, it is ok to have an annotated return type on object definitions. It is just ambiguous for handler definitions. In this method, you can see whether something is an object or handler definition by inspecting `continuationDetails`:

`None` -> object definition
`Some(ctx)` -> handler definition

We could report the error only in the latter case since only there the ambiguity arises.

Reply:
> We could report the error only in the latter case since only there the ambiguity arises.

Done in a18277b8b37d176e91ed4a268fabe921441e8277.

However, when done naively, this now introduces a problem since we don't check the user-supplied type.
<img width="488" alt="Screenshot 2022-09-23 at 15 01 57" src="https://user-images.githubusercontent.com/11269173/191966110-d33300f6-8f3e-49eb-b3a1-0302fd6d3c1a.png">

Ideally, I'd want to use something like `matchDeclared`:
https://github.com/effekt-lang/effekt/blob/b9b56fa12a0d4b2552818e9180182e38367aec43/effekt/shared/src/main/scala/effekt/Typer.scala#L1124

but right now it's overly specialised for `source.Param`s.

Moreover, I'm not actually sure that `matchDeclared` is the right thing to use here -- in a system without subtyping, this would just be an equality constraint:
https://github.com/effekt-lang/effekt/blob/a18277b8b37d176e91ed4a268fabe921441e8277/effekt/shared/src/main/scala/effekt/Typer.scala#L357-L369

Reply:
The important parts of this conversation are repeated and clarified below so I'm resolving this thread.